model/riscv_device_tree.sail

```diff
@@ -6,6 +6,156 @@
 /*  SPDX-License-Identifier: BSD-2-Clause                                                */
 /*=======================================================================================*/
 
+// Generators for Device-Tree source corresponding to the model
+// configuration.
+// NOTE: These assume a valid model configuration.
+
+// Generates the value for the "riscv,isa-base" property.
+function generate_isa_base() -> string = {
+  "rv" ^ dec_str(xlen) ^ "i"
+}
+
+/* Some notes from the "ISA Extension Naming Convections" section of
+   the manual:
+
+  . The first letter following the "Z" conventionally indicates the
+    most closely related alphabetical extension category,
+    IMAFDQLCBKJTPVH.  If multiple "Z" extensions are named, they
+    should be ordered first by category, then alphabetically within a
+    category—for example, "Zicsr_Zifencei_Ztso".
+
+  . All multi-letter extensions, including those with the "Z" prefix, must be
+    separated from other multi-letter extensions by an underscore, e.g.,
+    "RV32IMACZicsr_Zifencei".
+
+  . Standard supervisor-level extensions (prefixed with either 'Ss' or
+    'Sv') should be listed after standard unprivileged extensions, and
+    like other multi-letter extensions, must be separated from other
+    multi-letter extensions by an underscore. If multiple
+    supervisor-level extensions are listed, they should be ordered
+    alphabetically.
+
+  . Standard machine-level extensions (prefixed with the letters "Sm")
+    should be listed after standard lesser-privileged extensions, and
+    like other multi-letter extensions, must be separated from other
+    multi-letter extensions by an underscore. If multiple
+    machine-level extensions are listed, they should be ordered
+    alphabetically.
+
+  Once the Sail configuration supports extension version numbers, we could
+  add them to the generated isa string.
+
+  The device-tree specification for the ISA string is at:
+  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/riscv/extensions.yaml
+
+  As the above document uses lower-case, the generator below does too.
+*/
+
+// Generates the value for the "riscv,isa-extensions" property
+// (if `as_isa_ext == false`) or the "riscv,isa" property (otherwise).
+// NOTE: The "riscv,isa" property is marked deprecated but is probably
+// still widely used.
+
+function dt_wrap(s : string, as_isa_ext : bool) -> string = {
+  if   as_isa_ext
+  then ", \"" ^ s ^ "\""
+  // In "riscv,isa", single letter extensions are concatenated
+  // without a separator.
+  else if string_length(s) == 1
+  then s
+  else "_" ^ s
+}
+
+function generate_isa_string(as_isa_ext : bool) -> string = {
+  (if   as_isa_ext
+   then "\"i\""
+   else "rv" ^ dec_str(xlen) ^ "i")
+
+// First append the single letter extensions.
+
+^ (if hartSupports(Ext_M) then dt_wrap("m", as_isa_ext) else "")
```

Comment:
```suggestion
^ dt_wrap(hartSupports(Ext_M), "m", as_isa_ext)
```

How about passing a boolean as the first argument to dt_wrap  and push the if into that function? Then we can avoid all the if/else in this function?